miden-lib/src/notes/mod.rs

```diff
@@ -50,6 +57,27 @@ pub fn create_note(
             ProgramAst::from_bytes(p2idr_bytes).map_err(NoteError::NoteDeserializationError)?,
             vec![target.into(), recall_height.into(), ZERO, ZERO],
         ),
+        Script::ASWAP {
+            requested_asset,
```

Comment:
I think we need to add the `sender_id` here to compute the `RECIPIENT` correctly.

Reply:
Can't we get the sender_id in the masm using this: https://github.com/0xPolygonMiden/miden-base/blob/9806a922a3b99948f6254f8ae105904dd4eda7ea/miden-lib/asm/miden/sat/note.masm#L88-L101

Reply:
Yes, but here in that case you need the sender in Rust before you execute MASM to get the note input `RECIPIENT`.

Reply:
With the above discussion with @bobbinth It is not clear to me what we should do with the inputs to the ASWAP script.

For now it works with ASSET and RECIPIENT ( dummy ).

Should we pass ASSET and SERIAL_NUM or ASSET and RECIPIENT?

I have the feeling that the serial_num of the note is not known at the moment of the creation of the ASWAP script, only when the whole note is created. 

If we pass the ASSET and SERIAL_NUM then i will have to create my RECIPIENT in my masm code following these guidelines right?: hash(hash(hash(serial_num, [0; 4]), script_hash), input_hash)

Furthermore I am curious how it would work in practice. The wallet would produce the ASSET and SERIAL_NUM or RECIPIENT to be able to input the script then create a note and tx right?

Reply:
We can create `RECIPIENT` in Rust here - we have enough info to do it. To compute the recipient we need the following:

1. `script_root` - this would be root of `P2ID` script.
2. `input_hash` - this would be hash of inputs for the `P2ID` script from point 1. The inputs, in our case would be `[sender, 0, 0, 0]`.
3. `serial_num` - this could be passed in via `script` enum (see previous comment).

So, basically, we need a function which looks something like this:
```Rust
fn build_p2id_recipient(target: AccountId, serial_num: Word) -> Digest
```
Then, from here, you'd invoke this as:
```Rust
let recipient = build_p2id_recipient(sender, serial_num);
```

Reply:
From the wallet's standpoint, it would happen like so:
```Rust
let script = Script::ASWAP {
    asset: asset_i_want_to_receive,
    serial_num: serial_num_for_return_note
}

let note = create_note(
    script,
    asset_i_want_to_exchange,
    my_account_id,
    Some(atomic_swap_tag),
    serial_num_for_outgoing_note,
).unwrap();
```

One thing we haven't covered yet is how to compute a tag for an atomic swap. I have put some thoughts about it in https://github.com/0xPolygonMiden/miden-node/issues/104#issuecomment-1852377789 - but don't have a concrete proposal yet.

Reply:
In a following comment you said: > I think it is better to pass the tag via inputs. This way, the original sender controls all aspects of the returning note (e.g., they may want the returning note sent to a different address).

Hence I added `tag` as a note_script input enabling the user to manipulate their input easily. Is that what you were envisioning? 

```rust
    ASWAP {
        asset: Asset,
        serial_num: Word,
        tag: Felt,
    },
```

Reply:
Ah yes - you are correct. Sorry for the confusion!

Reply:
Or - actually, here we can skip the tag as we can set it to the sender's account ID. In MASM though, it would be an input.

The reasoning is that here we are being a bit more restrictive for simplicity purposes, but MASM should be more general. This way, we still have a standardized flexible MASM script, and if someone wants to use it in ways we don't support here, they can craft their notes manually.